tada.rb

```diff
@@ -0,0 +1,323 @@
+      @path.gsub('.tif', '.xml')
+    else
+      @metadata_file_path
+    end
+  end
+
+  def metadata()
+    # Pull needed fields out of the metadata file
+    begin
+      data = Nokogiri::XML(File.read(self.metadata_file))
+      {
+        'title' => data.xpath("//field[@name='title']//value//text()"),
+        'description' => data.xpath("//field[@name='description']//value//text()")
+      }
+    rescue
+      $logger.error "Fool, no metadata for #{self.tif_file}"
+      return nil
+    end
+  end
+
+  def ark()
+    if self.metadata != nil
+      # Try to figure out if there is already an ARK in the metadata file.
+      # If not, make one and add it to the metadata file.
+      @data = Nokogiri::XML(File.open(self.metadata_file))
+      ark_field =  @data.xpath("//field[@name='ark']//value").first
+      # Check to see if the ARK filed exists or is empty.
+      if ark_field.nil? || ark_field.text == ''
+
+        ark = create_ark()
+
+        # If there is no placeholder field, make the filed.
+        if ark_field.nil?
+          record = @data.xpath('//record//field')[-1]
+          record.add_next_sibling("<field name='ark'><value>#{ark}</value></field>")
+        # If there is one, populate it.
+        else
+          ark_field.content = ark
+        end
+        # Update the the metadata file.
+        File.open(self.metadata_file, 'w') do |updated|
+          updated << @data
+        end
+        # And return the ARK.
+        return ark
+      # Otherwise, just get the ARK from the metadata and return it.
+      else
+        ark_field.text
+      end
+    end
+  end
+
+  def create_ark()
+    # Private method to create an ARK via the pidman REST API.
+    pidman_auth = {
+      username: $config['pidman_user'],
+      password: $config['pidman_pass']
+    }
+
+    response = HTTParty.post \
+      'https://testpid.library.emory.edu/ark/', \
+      body: "domain=#{$config['pidman_domain']}&target_uri=myuri.org&name=#{self.metadata['title']}", \
+      basic_auth: pidman_auth
+    # The response will give us the full URL, we just want the PID.
+    if  "#{response.code}" == '201'
+      return response.body.split('/')[-1]
+    else
+      $logger.error "Failed to create ARK for #{self.file_name}. Response was #{response.code}"
+    end
+  end
+
+  # Make the create_ark method private.
+  private :create_ark
+
+end
+
+class GeoServer
+  # Class to provide access to the GeoServer.
+  def initialize()
+    $config = YAML.load_file('config.yaml')
+  end
+
+  # TODO make this work for shape files.
+  def endpoint()
+    # Return the proper endpoint to the GeoServer REST API.
+    "#{$config['geoserver_url']}/geoserver/rest/workspaces/" \
+    "#{$config['geoserver_workspace']}/coveragestores"
+  end
+
+  def auth()
+    # Return a hash for authenticaing to the REST API.
+    auth = {
+      username: $config['geoserver_user'],
+      password: $config['geoserver_pass']
+    }
+    return auth
+  end
+end
+
+def add_store(map)
```

Comment:
Would it make more sense for the this function as well as the add/update layer ones to be methods on the GeoServer class?